libarchive/archive.h

```diff
@@ -944,6 +944,9 @@ __LA_DECL int archive_write_disk_set_user_lookup(struct archive *,
     void (* /* cleanup */)(void *));
 __LA_DECL la_int64_t archive_write_disk_gid(struct archive *, const char *, la_int64_t);
 __LA_DECL la_int64_t archive_write_disk_uid(struct archive *, const char *, la_int64_t);
+#ifdef HAVE_COPY_FILE_RANGE
+__LA_DECL la_ssize_t archive_write_disk_copy_file(struct archive *, int, off_t);
```

Comment:
I like this new method, but if we're going to add it, it should make sense as a reusable service (not just as a cpio helper).
* What if the `fd_in` is not currently positioned at the beginning of the file?  What should happen?
* Should this function take an fd and size or should it take a filename instead?
* Clearly, we can provide it regardless of whether the platform supports this particular sys call.  (If other platforms provide similar services, we should plan to support them in the future as well.)
* The name could be more accurate:  It does not "copy a file"; it fills the current entry with the contents of the specified file.

Reply:
> I like this new method, but if we're going to add it, it should make sense as a reusable service (not just as a cpio helper).
> 
>   * What if the `fd_in` is not currently positioned at the beginning of the file?  What should happen?

As I said above, the function doesn't really care about that, it simply copies X bytes from one file to the other.  I guess that, if the final version of the API will still receive a file descriptor, this would need to be very well documented.

>   * Should this function take an fd and size or should it take a filename instead?

Again, as I said above, this would mean an extra syscall as cpio would already need that file open, right?  Should it be closed before?

>   * Clearly, we can provide it regardless of whether the platform supports this particular sys call.  (If other platforms provide similar services, we should plan to support them in the future as well.)

Right, moving the copy loop into the API as you suggest below would be a solution: the caller wouldn't need to bother knowing whether copy_file_range exists or not. 

>   * The name could be more accurate:  It does not "copy a file"; it fills the current entry with contents from the specified file.

Ok, I'll try to thing of something else.  Naming is hard :-)